interp/src/interpreter/steppers/group_interpreter.rs

```diff
@@ -0,0 +1,317 @@
+                    //value driving the port
+                    assigned_ports.insert(pa);
+                    //ok now proceed
+                    //the below (get) attempts to get from working_env HM first, then
+                    //backing_env Smoosher. What does it mean for the value to be in HM?
+                    //That it's a locked value?
+                    let old_val =
+                        self.state.get_from_port(&assignment.dst.borrow());
+                    let new_val_ref =
+                        self.state.get_from_port(&assignment.src.borrow());
+                    // no need to make updates if the value has not changed
+                    let port = assignment.dst.clone(); // Rc clone
+                    let new_val = new_val_ref.clone();
+
+                    if old_val != new_val_ref {
+                        updates_list.push((port, new_val)); //no point in rewriting same value to this list
+                        self.val_changed = Some(true);
+                    }
+                }
+            }
+
+            let assigned_const_ports: HashSet<_> = assigned_ports
+                .iter()
+                .map(PortAssignment::get_port)
+                .collect();
+
+            //now assign rest to 0
+            //first get all that need to be 0
+            for port in &possible_ports - &assigned_const_ports {
+                //need to set to zero, because unassigned
+                //ok now proceed
+
+                //need to find appropriate-sized 0, so just read
+                //width of old_val
+
+                let old_val = self.state.get_from_port(port);
+                let old_val_width = old_val.width(); //&assignment.dst.borrow().width()
+                let new_val = Value::from(0, old_val_width).unwrap();
+
+                if old_val.as_u64() != 0 {
+                    self.val_changed = Some(true);
+                }
+
+                //update directly
+                self.state.insert(port, new_val);
+            }
+
+            // perform all the updates
+            for (port, value) in updates_list {
+                self.state.insert(port, value);
+            }
+
+            let changed = eval_prims(&mut self.state, self.cells.iter(), false);
+            if changed {
+                self.val_changed = Some(true);
+            }
+        }
+    }
+    /// Advance the interpreter by a cycle, if possible
+    pub fn step(&mut self) {
+        self.step_cycle();
+        self.step_convergence();
+    }
+
+    /// Run interpreter until it is finished executing and return the output
+    /// environment
+    pub fn run_and_deconstruct(mut self) -> InterpreterState {
+        self.run();
+        self.deconstruct()
+    }
+
+    /// Run the interpreter until it finishes executing
+    pub fn run(&mut self) {
+        while !self.is_done() {
+            self.step();
+        }
+        self.step_convergence();
+    }
```

Comment:
Just have to say that I love how clear these "run" functions are… they make so much sense!